src/util/dateHelpers.js

```diff
@@ -39,26 +39,52 @@ export const numericToCalendar = (numDate) => {
 };
 
 /**
- * Convert a calendar date to a numeric one.
- * This function is meant to behave similarly to TreeTime's `numeric_date`
- * as found in v0.7*. Note that for negative dates, i.e. BCE, no fraction
- * in the year will be returned.
+ * Convert a YYYY-MM-DD string to a numeric date. This function is meant to
+ * behave similarly to TreeTime's `numeric_date` as found in v0.7*. For negative
+ * dates (i.e. BCE) we simply return the year (ignoring month / day). Ambiguity
+ * is optionally allowed in the form of YYYY-MM-XX or YYYY-XX-XX in which case
+ * the midpoint of the implied range is returned. All non compliant inputs
+ * return `undefined`.
  * @param {string} calDate in format YYYY-MM-DD
- * @returns {float} YYYY.F, where F is the fraction of the year passed
+ * @param {boolean} ambiguity
+ * @returns {float|undefined} YYYY.F, where F is the fraction of the year passed
  */
-export const calendarToNumeric = (calDate) => {
+export const calendarToNumeric = (calDate, ambiguity=false) => {
+  if (typeof calDate !== "string") return undefined;
   if (calDate[0]==='-') {
-    const pieces = calDate.substring(1).split('-');
-    return -parseFloat(pieces[0]);
+    const d = -parseFloat(calDate.substring(1).split('-')[0]);
+    return isNaN(d) ? undefined : d;
   }
-  /* Beware: for `Date`, months are 0-indexed, days are 1-indexed */
-  const [year, month, day] = calDate.split("-").map((n) => parseInt(n, 10));
+  const fields = calDate.split("-");
+  if (fields.length !== 3) return undefined;
+  const [year, month, day] = fields;
+  const [numYear, numMonth, numDay] = fields.map((d) => parseInt(d, 10));
+  
+  if (calDate.includes("X")) {
+    if (!ambiguity) return undefined
+    if (year.includes("X")) return undefined;
+    if (month.includes("X")) {
+      if (isNaN(numYear) || day!=="XX") return undefined
+      return numYear + 0.5;
+    }
+    if (isNaN(numYear) || isNaN(numMonth)) return undefined
+    const range = [
+      _yearMonthDayToNumeric(numYear, numMonth, 1),
+      _yearMonthDayToNumeric(numMonth===12?numYear+1:numYear, numMonth===12?1:numMonth+1, 1)
+    ]
+    return range[0] + (range[1]-range[0])/2
+  }
+  return _yearMonthDayToNumeric(numYear, numMonth, numDay)
```

Comment:
_non-blocking_

the logic would be a little more apparent here with some "else"s 

```suggestion
  if (calDate.includes("X")) {
    if (!ambiguity) return undefined
    else if (year.includes("X")) return undefined;
    else if (month.includes("X")) {
      if (isNaN(numYear) || day!=="XX") return undefined
      else return numYear + 0.5;
    }
    else if (isNaN(numYear) || isNaN(numMonth)) return undefined
    const range = [
      _yearMonthDayToNumeric(numYear, numMonth, 1),
      _yearMonthDayToNumeric(numMonth===12?numYear+1:numYear, numMonth===12?1:numMonth+1, 1)
    ]
    else return range[0] + (range[1]-range[0])/2
  }
  else return _yearMonthDayToNumeric(numYear, numMonth, numDay)
```

Reply:
ha! I actually don't like this style too much and have recommended the opposite to others 😄

Reply:
As we discussed earlier on our call, this is peak "take it or leave it" feedback. I like the explicitness that the `else`s introduce; it means I don't have to reason as much about what's a fall-through condition and what's an alternative branch.